README.md
Laravel GeoLocation Package
===========================

Laravel package to get the details about Region, City and Country for a given IP Address.

This package is a wrapper for [IpInfo](https://ipinfo.io) Provider, however we have plan to implement a driver for 
MaxMind GeoIP2 databases.

### Install

    composer require adrianorosa/laravel-geolocation

> This package supports the latest version of Laravel, for now 5.8+ was tested, but older versions should work fine.

As of Laravel 5.5 there is no need to add the ServiceProvider within the `AppServiceProvider` array. 
Laravel loads this provider using Package Discovery.

### Usage

There is no need to additional setup to start using, once you install it via composer you can call the Facade:

```php
<?php
use Adrianorosa\GeoLocation\GeoLocation;

$details = GeoLocation::lookup('8.8.8.8');

echo $details->getIp();
// 8.8.8.8

echo $details->getCity();
// Mountain View

echo $details->getRegion();
// California

echo $details->getCountry();
// United States

echo $details->getLatitude();
// 37.386

echo $details->getLongitude();
// -122.0838

var_dump($details->toArray());
// Array
// (
//  [city] => Mountain View
//  [region] => California
//  [country] => United States
//  [countryCode] => US
//  [latitude] => 37.386
//  [longitude] => -122.0838
//)
``` 

### Console Command

To display details about an Ip or Your current network Ip Address you may use the artisan command: 

```bash
php artisan geolocation:lookup --ip 8.8.8.8

+---------+---------------+------------+----------------+-------------+----------+-----------+
| ip      | city          | region     | country        | countryCode | latitude | longitude |
+---------+---------------+------------+----------------+-------------+----------+-----------+
| 8.8.8.8 | Mountain View | California | Estados Unidos | US          | 37.386   | -122.0838 |
+---------+---------------+------------+----------------+-------------+----------+-----------+
```

## Override Configuration and translations

This package comes with a little configuration for IpInfo and translations for Country Names.

You can stick with default values, which works fine, or publish using the following commands:

Publish all config and translations 

    php artisan vendor:publish

Publish config

    php artisan vendor:publish --tag=geolocation-config
    
Publish translations

    php artisan vendor:publish --tag=geolocation-translations
    
    
## Author

**Adriano Rosa** (https://adrianorosa.com)  

## Licence

Copyright © 2021, Adriano Rosa  <info@adrianorosa.com>
All rights reserved.

For the full copyright and license information, please view the LICENSE 
file that was distributed within the source root of this project.


CHANGELOG.md
CHANGELOG
=========

### v1.2.0

- 2023-03-07 - update version constraint for require-dev testbench package
- 2023-03-07 - typo fix
- 2023-03-07 - Merge pull request #6 from BrekiTomasson/dev-l10-compatibility

### v1.1.0

- 2022-04-30 - update Travis config to support PHP `8.0` and `8.1`.
- 2022-04-27 - added support for Laravel `^9.0`.

### v1.0.0

- 2021-01-25 - fix composer requirements
- 2021-01-25 - add return type declaration
- 2021-01-25 - update README
- 2021-01-25 - removed unused class
- 2021-01-25 - upgrade testbench to 6.9 and guzzle to 7.0
- 2021-01-25 - upgrade tests

### v0.3.0

- 2019-11-15  - update requirement version constrain for laravel 7.0 and 8.0

### v0.2.0

- 2019-11-15  - update requirement version constrain for laravel 6.*

### v0.1.3

- 2019-08-29 - Removed DeferrableProvider interface, now the package is not deferred by default
- 2019-08-29 - Add new Facade `GeoLocation::countries` method to get the translation list of countries codes

### v0.1.2

 - 2019-08-28 - Add TravisCI
 - 2019-08-28 - Add new method GeoLocationDetails::getCountryCode()
 
 ### v0.1.1

 - 2019-08-22 - Add support for laravel 6.0
 
### v0.1.0

 - 2019-08-13 - Initial Release


src/GeoLocationDetails.php
<?php

namespace Adrianorosa\GeoLocation;

use Illuminate\Contracts\Support\Arrayable;

/**
 * Class GeoLocationDetails.
 *
 * @author Adriano Rosa <https://adrianorosa.com>
 * @date 2019-08-13 13:24
 *
 * @package Adrianorosa\GeoLocation
 */
class GeoLocationDetails implements \JsonSerializable, Arrayable
{
    /**
     * @var string
     */
    protected $ip;

    /**
     * @var string
     */
    protected $city;

    /**
     * @var string
     */
    protected $region;

    /**
     * @var string
     */
    protected $country;

    /**
     * @var string
     */
    protected $countryCode;

    /**
     * @var float
     */
    protected $latitude;

    /**
     * @var float
     */
    protected $longitude;

    /**
     * GeoLocationDetails constructor.
     *
     * @param  array $data
     */
    public function __construct($data = [])
    {
        $this->parse($data);
    }

    /**
     * Get the IP Address.
     *
     * @return string|null
     */
    public function getIp()
    {
        return $this->ip;
    }

    /**
     * Get the City name.
     *
     * @return string|null
     */
    public function getCity()
    {
        return $this->city;
    }

    /**
     * Get teh region name.
     *
     * @return string|null
     */
    public function getRegion()
    {
        return $this->region;
    }

    /**
     * Get the country name.
     *
     * @return string|null
     */
    public function getCountry()
    {
        return $this->country;
    }

    /**
     * Get the country ISO Code.
     *
     * @return string|null
     */
    public function getCountryCode()
    {
        return $this->countryCode;
    }

    /**
     * Get the Latitude value.
     *
     * @return float|null
     */
    public function getLatitude()
    {
        return $this->latitude;
    }

    /**
     * Get the Longitude value.
     *
     * @return float|null
     */
    public function getLongitude()
    {
        return $this->longitude;
    }

    /**
     * Parses the raw response.
     *
     * @param  mixed $data
     */
    protected function parse($data)
    {
        if (is_string($data)) {
            $data = json_decode($data, true);
        }

        foreach ((array) $data as $key => $value) {
            if (property_exists($this, $key)) {
                if ($key === 'country') {
                    $this->countryCode = $value;
                    $this->{$key} = $this->formatCountry($value);
                } else {
                    $this->{$key} = $value;
                }
            }

            if ($key === 'loc') {
                $this->formatCoordinates($value);
            }
        }
    }

    /**
     * Parse the coordinate values into latitude and longitude.
     *
     * @param  string $value
     */
    protected function formatCoordinates($value): void
    {
        [$latitude, $longitude] = explode(',', $value);

        $this->latitude = (float) $latitude;
        $this->longitude = (float) $longitude;
    }

    /**
     * Format the country name.
     *
     * @param  string $countryCode
     *
     * @return mixed
     */
    protected function formatCountry($countryCode)
    {
        $countries = trans('geolocation::countries');

        return array_key_exists($countryCode, $countries) ? $countries[$countryCode] : $countryCode;
    }

    /**
     * Get list of location items as an array.
     *
     * @return array
     */
    public function toArray()
    {
        return [
            'city' => $this->city,
            'region' => $this->region,
            'country' => $this->country,
            'countryCode' => $this->countryCode,
            'latitude' => $this->latitude,
            'longitude' => $this->longitude,
        ];
    }

    /**
     * {@inheritdoc}
     */
    public function jsonSerialize()
    {
        return $this->toArray();
    }
}
